src/metabase/fingerprinting.clj

```diff
@@ -0,0 +1,452 @@
+(ns metabase.fingerprinting
+  "Fingerprinting (feature extraction) for various models."
+  (:require [bigml.histogram.core :as hist]
+            [bigml.sketchy.hyper-loglog :as hyper-loglog]
+            (clj-time [coerce :as t.coerce]
+                      [core :as t]
+                      [format :as t.format]
+                      [periodic :as t.periodic])
+            [kixi.stats.math :as math]
+            [kixi.stats.core :as stats]
+            [metabase.db.metadata-queries :as metadata]
+            (metabase.models [card :refer [Card]]
+                             [field :refer [Field]]
+                             [segment :refer [Segment]]
+                             [table :refer [Table]])
+            [redux.core :as redux]
+            [tide.core :as tide]))
+
+(def ^:private ^:const percentiles (range 0 1 0.1))
+
+(defn histogram
+  "Transducer that summarizes numerical data with a histogram."
+  ([] (hist/create))
+  ([acc] acc)
+  ([acc x] (hist/insert! acc x)))
+
+(defn histogram-categorical
+  "Transducer that summarizes categorical data with a histogram."
+  ([] (hist/create))
+  ([acc] acc)
+  ([acc x] (hist/insert-categorical! acc (when x 1) x)))
+
+(defn rollup
+  "Transducer that groups by `groupfn` and reduces each group with `f`.
+   Note the contructor airity of `f` needs to be free of side effects."
+  [f groupfn]
+  (let [init (f)]
+    (fn
+      ([] (transient {}))
+      ([acc]
+       (into {}
+         (map (fn [[k v]]
+                [k (f v)]))
+         (persistent! acc)))
+      ([acc x]
+       (let [k (groupfn x)]
+         (assoc! acc k (f (get acc k init) x)))))))
+
+(defn safe-divide
+  "Like `clojure.core//`, but returns nil if denominator is 0."
+  [numerator & denominators]
+  (when (or (and (not-empty denominators) (not-any? zero? denominators))
+            (and (not (zero? numerator)) (empty? denominators)))
+    (apply / numerator denominators)))
+
+(defn growth
+  "Relative difference between `x1` an `x2`."
+  [x2 x1]
+  (when (every? some? [x2 x1])
+    (safe-divide (* (if (neg? x1) -1 1) (- x2 x1)) x1)))
+
+(defn bins
+  "Return centers of bins and thier frequencies of a given histogram."
+  [histogram]
+  (let [bins (hist/bins histogram)]
+    (or (some->> bins first :target :counts (into {}))
+        (into {}
+          (map (juxt :mean :count))
+          bins))))
+
+(def ^{:arglist '([histogram])} nil-count
```

Comment:
If you're ever in doubt as to which it's supposed to be just look at the metadata for a function. e.g.

```clojure
(meta #'bins)
```

in this namespace in your REPL